### PR TITLE
keycloack token refreshing added + logout on token expiry

### PR DESCRIPTION
**Pull request by dudiiiiiiii**

### CODE REVIEW

File: src/services/AuthService.tsx [hunk 1]
```diff
@@ -284,7 +290,8 @@ const Auth = ({ children }: GenericContextProviderProps) => {
 	);
 
 	const tokenRefresh = useCallback(() => {
-		if (!initialized || user?.source !== 'keycloak') return Promise.resolve();
+		if (!initialized || !keycloak.authenticated || user?.source != 'keycloak')
```
Comment (grzanka) [inline src/services/AuthService.tsx:293]: comment is needed to explain the logic

File: src/services/AuthService.tsx [hunk 2]
```diff
@@ -295,12 +302,16 @@ const Auth = ({ children }: GenericContextProviderProps) => {
...
-			.catch(reason => {});
+			.catch(reason => {
+				logout();
+			});
 	}, [initialized, user?.source, keycloak, enqueueSnackbar]);
 
 	useIntervalAsync(
 		tokenRefresh,
-		initialized && keycloak.authenticated ? keyCloakInterval : undefined
+		initialized && keycloak.authenticated && keycloak.isTokenExpired()
```
Comment (grzanka) [inline src/services/AuthService.tsx:312]: comment is needed to explain the logic

File: docs/authentication.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        end
+    end
+    opt user doesn't have access
+        AuthService ->> User: Message with access denied
+    end
+    loop Refresh backend connection every 3 minutes
+        AuthService ->> Backend: Refresh token (GET auth/refresh)
+        Backend ->> AuthService: Response with new backend access token in cookies
+    end
+    loop Refresh token every 1/3 of tokens lifetime
```
Comment (grzanka) [inline docs/authentication.md:40]: add what is a typical token lifetime (in minutes)

**Comment by grzanka** [comment]
Token seems valid 30 min:

image

After 30 min I click on "Simulation" tab and got "No simulations found". On list of queries to backend I get:

image

Moreover I am still logged in, despite expired token.

**Comment by grzanka** [comment]
@dudiiiiiiii please rebase and make some tests on yap-dev

File: src/services/AuthService.tsx (hunk 2 above)
Comment (dudiiiiiiii) [reply src/services/AuthService.tsx:309]: previous logic correct no chages

File: src/services/AuthService.tsx (hunk 1 above)
Comment (dudiiiiiiii) [reply src/services/AuthService.tsx]: previous logic correct no chages

**Comment by grzanka** [comment]
@dudiiiiiiii can you synchronise this branch with recent changes on master ?

Converter is there at 3189a07
https://github.com/yaptide/ui/tree/master/src/libs

**Comment by dudiiiiiiii** [comment]
> Member

I rebased to newest master un yaptide/ui.
From what i can see converter at - 3189a07 is the newest version

**Comment by grzanka** [comment]
> > @grzanka
> 
> I rebased to newest master un yaptide/ui. From what i can see converter at - 3189a07 is the newest version

Check here:
https://github.com/yaptide/ui/tree/1594-user-still-logged-in-when-browser-reopened-after-long-time-not-used
Screenshot 2024-10-15 at 07 57 04

It still says "16 commits behind master"

**Comment by grzanka** [comment]
To avoid such lengthy commit history, could you squash merge your changes into single commit and put that on top of the newest master ?

**Comment by dudiiiiiiii** [comment]
> To avoid such lengthy commit history, could you squash merge your changes into single commit and put that on top of the newest master ?

Current rebase should be good (previously i did something wrong) and commit history is fixed

File: src/services/AuthService.tsx [hunk 4]
```diff
@@ -194,7 +194,13 @@ const Auth = ({ children }: GenericContextProviderProps) => {
 				);
 		};
 
-		if (!initialized || !keycloak.authenticated) return;
+		if (!initialized || !keycloak.authenticated) {
```
Comment (grzanka) [inline src/services/AuthService.tsx:197]: Here is a table documenting what happens in the code:

| initialized | keycloak.authenticated | !initialized \|\| !keycloak.authenticated | initialized && !keycloak.authenticated | logout() | return |
|-------------|------------------------|-------------------------------------------|----------------------------------------|----------|--------|
| false       | false                   | true                                      | false                                  | no       | yes    |
| false       | true                    | true                                      | false                                  | no       | yes    |
| true        | false                   | true                                      | true                                   | yes      | yes    |
| true        | true                    | false                                     | false                                  | no       | no     |

File: src/services/AuthService.tsx [hunk 5]
```diff
@@ -194,7 +194,13 @@ const Auth = ({ children }: GenericContextProviderProps) => {
...
+			if (initialized && !keycloak.authenticated) {
+				logout();
+			}
+
+			return;
+		}
```
Comment (grzanka) [inline src/services/AuthService.tsx:203]:
```suggestion
		if (!initialized) {
		        // keycloak authentication system no initilized (working in demo mode or keycloak not available?)
		        // skipping futher checks (for example: to ask backend if the keycloak token is OK
			return;
		} else {
		        // keycloak authentication is initilized, so it makes sense to check if user is authenticated in keycloak
		        if(!keycloak.authenticated) {
		                // user not authenticated, forcing logout from yaptide app
				logout();		        
		               // skipping futher checks (for example: to ask backend if the keycloak token is OK
		         	return;
		        }
		        // user authenticated, we proceed with further checks
		
		}

		
						if (!initialized || !keycloak.authenticated) {
			if (initialized && !keycloak.authenticated) {
				logout();
			}

			return;
		}
```

Comment (grzanka) [reply src/services/AuthService.tsx:203]: I checked the double logic and it took me really long while do deal with negations, or and other boolean stuff. The code is working properly, but its very far from being readable.

What do you think of sacrificing the compactness of the code for the readability ?

What about an explicit version with few comments as I suggest above ?

Comment (dudiiiiiiii) [reply src/services/AuthService.tsx:203]: Comments are clear and correct. This way is much more readable. I'll add the changes tomorrow.

Comment (grzanka) [reply src/services/AuthService.tsx:203]: Then I will make some tests again. I was not sure about description of what does `initialized` means, maybe you know better

**Comment by grzanka** [comment]
I am doing some tests on yap-dev now, once I'm done I will give an approve

**Review by grzanka — APPROVED** [review]

**Pull request merged**